src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs

```diff
@@ -453,10 +517,59 @@ public object VisitTable(TableExpression tableExpression, SearchOptions context)
                         _cteToLimit.Add(_tableExpressionCounter);
                     }
 
+                    // Update target reference cte dictionary
+                    // var curCte = TableExpressionName(_tableExpressionCounter);
+                    var curLimitCte = TableExpressionName(_tableExpressionCounter + 1);
+
+                    // Add current cte limit to the dictionary
+                    if (includeExpression.Reversed)
+                    {
+                        AddIncludeLimitCte(includeExpression.SourceResourceType, curLimitCte);
+                    }
+                    else
+                    {
+                        // Not reversed and a specific target type is provided as the 3rd part of include value
+                        if (includeExpression.TargetResourceType != null)
+                        {
+                            AddIncludeLimitCte(includeExpression.TargetResourceType, curLimitCte);
+                        }
+                        else if (includeExpression.ReferenceSearchParameter != null)
+                        {
+                            includeExpression.ReferenceSearchParameter.TargetResourceTypes?.ToList().ForEach(t => AddIncludeLimitCte(t, curLimitCte));
+                        }
+                    }
+
+                    // Handle Multiple Results sets to include from
+                    if (_includeFromCtes?.Count > 1 && _curFromCteIndex >= 0 && _curFromCteIndex < _includeFromCtes.Count - 1)
+                    {
+                        StringBuilder.Append($"),{Environment.NewLine}");
+
+                        // If it's not the last result set, append a new IncludeLimit cte, since IncludeLimitCte was not created for the current cte
```

Comment:
why if its the last you dont add another limit? because anyway at the end we add one?

Reply:
Because an IncludeLimit expression is added after every IncludeExpression in IncludeReqriter